upb/def.c

```diff
@@ -2030,13 +2049,148 @@ static void create_oneofdef(
+  }
+  ch = upb_ascii_lower(ch);
+  if (ch >= 'a' && ch <= 'f') {
+    return ch - 'a' + 0xa;
+  }
+  *src -= 1;  // Char wasn't actually a hex digit.
+  return -1;
+}
+
+static char upb_symtab_ParseHexEscape(symtab_addctx* ctx, const upb_fielddef* f,
+                                      const char** src, const char* end) {
+  char hex_digit = upb_symtab_TryGetHexDigit(ctx, f, src, end);
+  if (hex_digit < 0) {
+    symtab_errf(ctx,
+                "\\x cannot be followed by non-hex digit in field '%s' default",
+                upb_fielddef_fullname(f));
+    return 0;
+  }
+  unsigned int ret = hex_digit;
+  while ((hex_digit = upb_symtab_TryGetHexDigit(ctx, f, src, end)) >= 0) {
+    ret = (ret << 4) | hex_digit;
+  }
+  if (ret > 0xff) {
+    symtab_errf(ctx, "Value of hex escape in field %s exceeds 8 bits",
+                upb_fielddef_fullname(f));
+    return 0;
+  }
+  return ret;
+}
+
+char upb_symtab_TryGetOctalDigit(const char** src, const char* end) {
+  char ch;
+  if (!upb_symtab_TryGetChar(src, end, &ch)) return -1;
+  if (ch >= '0' && ch <= '7') {
+    return ch - '0';
+  }
+  *src -= 1;  // Char wasn't actually a octal digit.
+  return -1;
+}
+
+static char upb_symtab_ParseOctalEscape(symtab_addctx* ctx,
+                                        const upb_fielddef* f, char first,
+                                        const char** src, const char* end) {
+  char ch = first - '0';
+  char digit;
+  if ((digit = upb_symtab_TryGetOctalDigit(src, end)) >= 0) {
+    ch = (ch << 3) | digit;
+  }
+  if ((digit = upb_symtab_TryGetOctalDigit(src, end)) >= 0) {
+    ch = (ch << 3) | digit;
+  }
+  return ch;
+}
+
+static char upb_symtab_ParseEscape(symtab_addctx* ctx, const upb_fielddef* f,
+                                   const char** src, const char* end) {
+  char ch;
+  if (!upb_symtab_TryGetChar(src, end, &ch)) {
+    symtab_errf(ctx, "unterminated escape sequence in field %s",
+                upb_fielddef_fullname(f));
+    return 0;
+  }
+  switch (ch) {
+    case 'a':
+      return '\a';
+    case 'b':
+      return '\b';
+    case 'f':
+      return '\f';
+    case 'n':
+      return '\n';
+    case 'r':
+      return '\r';
+    case 't':
+      return '\t';
+    case 'v':
+      return '\v';
+    case '\\':
+      return '\\';
+    case '\'':
+      return '\'';
+    case '\"':
+      return '\"';
+    case '?':
+      return '\?';
+    case 'x':
+    case 'X':
+      return upb_symtab_ParseHexEscape(ctx, f, src, end);
+    case '0':
+    case '1':
+    case '2':
+    case '3':
+    case '4':
+    case '5':
+    case '6':
+    case '7':
+      return upb_symtab_ParseOctalEscape(ctx, f, ch, src, end);
+  }
+  symtab_errf(ctx, "Unknown escape sequence: \\%c", ch);
+}
+
+static str_t* unescape(symtab_addctx* ctx, const upb_fielddef* f,
+                      const char* data, size_t len) {
+  // Size here is an upper bound; escape sequences could ultimately shrink it.
+  str_t* ret = symtab_alloc(ctx, sizeof(*ret) + len);
```

Comment:
this is another case where the "give back data" part of an arena could come in handy

Reply:
Acknowledged.